internal/auth/providers/okta.go

```diff
@@ -0,0 +1,416 @@
+package providers
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/buzzfeed/sso/internal/auth/circuit"
+	"github.com/buzzfeed/sso/internal/pkg/groups"
+	log "github.com/buzzfeed/sso/internal/pkg/logging"
+	"github.com/buzzfeed/sso/internal/pkg/sessions"
+	"github.com/datadog/datadog-go/statsd"
+)
+
+// OktaProvider is an implementation of the Provider interface.
+type OktaProvider struct {
+	*ProviderData
+	StatsdClient *statsd.Client
+	cb           *circuit.Breaker
+	GroupsCache  groups.MemberSetCache
+}
+
+type UserInfoResponse struct {
+	EmailAddress  string   `json:"email"`
+	EmailVerified bool     `json:"email_verified"`
+	Groups        []string `json:"groups"`
+}
+
+// NewOktaProvider returns a new OktaProvider and sets the provider url endpoints.
+func NewOktaProvider(p *ProviderData, orgName string) (*OktaProvider, error) {
+	if orgName == "" {
+		return nil, errors.New("missing setting: okta-org-name")
+	}
+	p.ProviderName = "Okta"
+	// interact with resource owner and obtain an authorization grant
+	if p.SignInURL.String() == "" {
+		p.SignInURL = &url.URL{
+			Scheme: "https",
+			Host:   fmt.Sprintf("%s.oktapreview.com", strings.Trim(orgName, "\"")),
+			Path:   "/oauth2/default/v1/authorize"}
+		// to get a refresh token. see https://developer.okta.com/authentication-guide/tokens/refreshing-tokens/
+	}
+	// returns access tokens, ID tokens or refresh tokens
+	if p.RedeemURL.String() == "" {
+		p.RedeemURL = &url.URL{
+			Scheme: "https",
+			Host:   fmt.Sprintf("%s.oktapreview.com", strings.Trim(orgName, "\"")),
+			Path:   "/oauth2/default/v1/token"}
```

Comment:
I wonder if we can also generic-ify these paths, they could vary based on the namely structure of the authentication server on the okta side, right? i.e. it could be a non-`default` authentication service?

Reply:
Yup. `default` may be replaced with the ID of the auth server. I'll modify to allow this!